Pass FieldLevelEncryptionId as a parameter for all cache behaviors

willthames opened the pull request:
##### SUMMARY

FieldLevelEncryptionId is a mandatory parameter for UpdateDistribution.
We can't delete with an UpdateDistribution, and it's difficult to not
set FieldLevelEncryptionId just for CreateDistribution.

As a result cloudfront_distribution now requires 1.9.22

Ensure botocore >= 1.10 gets installed through installing boto3 >= 1.7
otherwise tests fail!

Update cloudfront_distribution tests to remove references to
test_identifier so test suite actually works

Fixes #40724

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
cloudfront_distribution

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```
ansible 2.7.0.dev0 (devel c4a6bce69f) last updated 2018/06/21 14:22:31 (GMT +1000)
  config file = None
  configured module search path = [u'/Users/will/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/will/src/ansible/lib/ansible
  executable location = /Users/will/src/ansible/bin/ansible
  python version = 2.7.14 (default, Mar  9 2018, 23:57:12) [GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.39.2)]
```


##### ADDITIONAL INFORMATION
Tested using cloudfront_distribution test suite

pierreant-p commented:
I think this should go in 2.5.6 and not wait for 2.6 or 2.7 because the cloudfront module is pretty much broken and not usable right now.

willthames commented:
Candidate for 2.5 and 2.6 backport

mattclay commented on test/runner/requirements/integration.cloud.aws.txt:
```diff
@@ -1,2 +1,2 @@
 boto
-boto3
+boto3 >= 1.7
```
Constraints should be in `constraints.txt`, with a comment explaining why they are needed.

willthames replied:
Oh, I thought I'd fixed that, I must not have pushed the fix properly. Thanks for the explanation though.

s-hertel commented on test/runner/requirements/constraints.txt:
```diff
@@ -17,3 +17,4 @@ ntlm-auth >= 1.0.6 # message encryption support
 requests-ntlm >= 1.1.0 # message encryption support
 requests-credssp >= 0.1.0 # message encryption support
 voluptuous >= 0.11.0 # Schema recursion via Self
+boto >= 1.7 # Causes botocore >= 1.10 which fixes cloudfront_distribution
```
Should that be boto3? Or should it just assert the right minimum of botocore is installed to be on the safe side, since it's possible for someone to install an earlier botocore on a later boto3.

willthames replied:
Oh, I meant to delete this, I no longer require it.

s-hertel commented on lib/ansible/modules/cloud/amazon/cloudfront_distribution.py:
```diff
@@ -1864,6 +1875,10 @@ def main():
             ['default_origin_domain_name', 'alias'],
         ]
     )
+    if not module.botocore_at_least('1.9.22'):
```
Should this become mandatory for creating and deleting a distribution? Or just for updating one? It doesn't look like AWS fails without it for creating/deleting.

willthames replied:
I do address this in the PR description. It's not possible to delete a distribution without updating it first to disable it. It would be possible to create one, but the logic of creation/update is so intertwined that it's easier to just enforce this.

s-hertel commented on lib/ansible/modules/cloud/amazon/cloudfront_distribution.py:
```diff
@@ -180,6 +182,7 @@
             I(max_ttl)
             I(compress)
             I(lambda_function_associations[])
+            I(field_level_encryption_id)
```
Should there be a note in the docs that this is required now?

willthames replied:
It's not actually required. You can set it if you choose, otherwise we just default it to the empty string (which would ideally be what AWS do, but they don't).

s-hertel replied:
Ah, I see. :-)

s-hertel reviewed (commented):
Looks good, I've tested that this resolves the problem. I just have some uncertainty about backwards compat where possible.

willthames commented:
Please don't merge before an alternative approach has been investigated - https://github.com/ansible/ansible/commit/501a8c4d78af5ab5369350b96b996b487dc00a17

willthames committed:
Handle FieldLevelEncryptionId in cloudfront_distribution

Ensure that FieldLevelEncryptionId is properly handled - passing it if
set, and keeping it if returned by GetDistribution

Update cloudfront_distribution tests to remove references to
test_identifier so test suite actually works

Fixes #40724

willthames commented:
@s-hertel this is now a much simpler fixed and all of your backward compatibility concerns should be alleviated.

ryansb commented:
rebuild_merge
